### DISCUSSION TITLE
How to restart trial of given id

### DISCUSSION THREAD

**Discussion by jakubMitura14** [optuna/optuna#5102] [Q&A] [answered]
Hello I want to start a particular trial I get it by
```
frozen=storage.get_trial(int(old_trial_id))
```

and then I want to run it and the result be saved to SQL for parallel tuning. I tried 
```
study.add_trial(frozen)
study.optimize(objective, n_trials=900,gc_after_trial=True)

```

but it starts a trial with different id then I added - what am I doing wrong?

**Comment by nzw0301** [comment]
Let me ask one question for clarification. Do you mean re-evaluating `frozen` from scratch or re-use intermediate values too?

**Reply by jakubMitura14** [reply]
reusing - this is used in case when machine stuck and I need to restart it, I am saving the trial id of each started and not finished trial and associate pytorch checkpoint with it. After restarting machine I want to load trial using trial id and checkpoint from associated path (I store this association locally on each machine)

**Reply by nzw0301** [reply]
Thanks. What's the trial state?

**Reply by nzw0301** [reply]
I thought https://optuna.readthedocs.io/en/stable/faq.html#can-i-monitor-trials-and-make-them-failed-automatically-when-they-are-killed-unexpectedly is relevant, but I'm not sure how to use this (the example does not work as I expected on my machine).

**Reply by jakubMitura14** [reply]
Frankly I do not know what state it is - and there is a risk that I have failed trial on other machine (the same SQL database) so it could lead to improper association with checkpoint and restarted trial.

Alternatively is it possible to create a new trial with state loaded from database/locally and run this one?

**Reply by nzw0301** [reply, marked as answer]
> Alternatively is it possible to create a new trial with state loaded from database/locally and run this one?

Yes! 

```python
import optuna

failed_trial_number = 2

# buggy objective.
def obj(t: optuna.Trial):
    x = t.suggest_int("x", -1000, 1000)
    if t.number == failed_trial_number:
        return None
    return x
    
study = optuna.create_study()

study.optimize(obj, n_trials=5)


failed_trial = study.get_trials()[failed_trial_number]
study.enqueue_trial(failed_trial.params)

def new_obj(t: optuna.Trial):
    # load checkpoint, somehow
    x = t.suggest_int("x", -1000, 1000)
    return x

study.optimize(new_obj, n_trials=1)
trials = study.get_trials()
assert trials[failed_trial_number].params == trials[-1].params
assert trials[failed_trial_number].state != trials[-1].state
```

https://optuna.readthedocs.io/en/stable/reference/generated/optuna.study.Study.html#optuna.study.Study.enqueue_trial

Unfortunately, Intermediate values are not inherited here, so we need to copy it manually...

**Reply by jakubMitura14** [reply]
I will test it later today but it seems like a solution, thanks!!
